### PR TITLE
[WIP] Indentation sensitivity

### PR DESCRIPTION
**Pull request by robrix**
Implements indentation-sensitive parsing atop #1.
- [x] Parse comments and blank lines
- [x] Groups should only contain inline sequences, not nested elements
- [x] Parse indentation
- [x] Parse block elements with `>>-`

### CODE REVIEW

**Comment by jspahrsummers** [comment]
:heart:

**Comment by robrix** [comment]
Okay, we’re a bunch closer—mostly on the inline elements.

**Comment by jspahrsummers** [comment]
:sparkling_heart:

**Comment by robrix** [comment]
Down to 9 failures (and with more tests added, even).

**Comment by robrix** [comment]
(High water mark was 13 just a moment ago.)

**Comment by robrix** [comment]
We have a real start at parsing indented children. We’re getting mis-parses now, instead of just failures to parse :sparkles:

**Comment by robrix** [comment]
@jspahrsummers I’m calling it a night but feel free to poke at and/or review this.

File: OGDL/Parser.swift [hunk 1]
```diff
@@ -76,16 +84,98 @@ private func buildHierarchy(values: [String]) -> Node? {
...
+private let subgraph: Int -> Parser<[Node]>.Function = { n in
+	(descendents ++ lines(n + 1) --> { [ $0.byAppendingChildren($1) ] }) | adjacent
+}
+
+private let line: Int -> Parser<[Node]>.Function = fix { line in
+	{ n in
+		// fixme: block parsing: ignore(%char_space+ ++ block(n))|?) ++
+		indentation(n) >>- { n in
+			subgraph(n) ++ optionalSpace ++ (comment | br)+
+		}
```
Comment (robrix) [inline OGDL/Parser.swift:174]: :boom: :camel:

File: OGDL/Parser.swift [hunk 2]
```diff
@@ -76,16 +84,98 @@ private func buildHierarchy(values: [String]) -> Node? {
...
+private func interleave<T, U>(separator: Parser<U>.Function, parser: Parser<T>.Function) -> Parser<[T]>.Function {
+	return (parser ++ (ignore(separator) ++ parser)*) --> { [$0] + $1 }
+}
+
+private func foldr<S: SequenceType, Result>(sequence: S, initial: Result, combine: (S.Generator.Element, Result) -> Result) -> Result {
+	var generator = sequence.generate()
+	return foldr(&generator, initial, combine)
+}
+
+private func foldr<G: GeneratorType, Result>(inout generator: G, initial: Result, combine: (G.Element, Result) -> Result) -> Result {
```
Comment (robrix) [inline OGDL/Parser.swift:109]: I wrote this before I saw the extant build hierarchy function :disappointed:

**Comment by jspahrsummers** [comment]
Down to 4 failures.

At this point, the only remaining failing tests are comma-separated siblings and parenthesized groups (AFAICT). I think I'd like to merge this in, clean it up a bit, then start using it within Carthage (because the functionality we need is basically complete now)—though there's no need to stop working on it separately.

@robrix What do you think?

**Comment by robrix** [comment]
Nice one! :sparkles:

That sounds great to me :+1:

**Pull request merged**